File: intelligence/deep_wonders/src/nn.cc
// deep_wonders:v1
#include "nn.h"

#include <assert.h>
#include <stdlib.h>

namespace deep_wonders {

NN *
nn_new( void )
{
        NN *nn = (NN *)calloc( 1, sizeof( *nn ) );
        assert( nn != NULL );
        return nn;
}

void
nn_free( NN *nn )
{
        if ( nn == NULL ) return;
        free( nn );
}

void
nn_evaluate( NN * /*nn*/, Game *g, float *policy_out, float *value_out )
{
        /* Uniform policy over legal actions only. */
        int actions[NUM_ACTIONS];
        int n = g->LegalActions( actions );
        for ( int i = 0; i < NUM_ACTIONS; i++ ) policy_out[i] = 0.0f;
        for ( int i = 0; i < n; i++ ) policy_out[actions[i]] = 1.0f / (float)n;
        *value_out = 0.0f;
}

}  // namespace deep_wonders

File: intelligence/deep_wonders/src/nn.rs
use crate::game::{Game, NUM_ACTIONS};

pub struct Nn {
    _placeholder: i32,
}

impl Nn {
    pub fn new() -> Nn {
        Nn { _placeholder: 0 }
    }

    /// Evaluate the game state. Returns uniform policy over legal actions and value=0.
    pub fn evaluate(&self, g: &Game) -> ([f32; NUM_ACTIONS], f32) {
        let _obs = g.observe(); // will feed into network; unused in stub
        let actions = g.legal_actions();
        let mut policy = [0.0f32; NUM_ACTIONS];
        let n = actions.len() as f32;
        for &a in &actions {
            policy[a.encode()] = 1.0 / n;
        }
        (policy, 0.0)
    }
}
